File: java/src/main/java/deltix/luminary/ProjectSaver.java
package deltix.luminary;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.OutputStreamWriter;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;

public class ProjectSaver {
    public static void save(ProjectDef projectDef, String rootDirectory) throws IOException {
        if (projectDef.getFileName() == null)
            throw new IllegalArgumentException("Built-in projects cannot be exported.");
        final Path directory = Paths.get(rootDirectory);
        if (!Files.exists(directory))
            throw new IllegalArgumentException("Directory does not exist.");

        saveProjectFile(projectDef, directory);
        for (NamespaceDef namespaceDef : projectDef.getNamespaces().values())
            saveNamespace(namespaceDef, directory);

    }

    private static void saveProjectFile(ProjectDef projectDef, Path directory) throws IOException {
        final ProjectFile projectFile = new ProjectFile();
        projectFile.References = projectDef.getReferences().keySet().toArray(new String[0]);
        projectFile.Sources = new ProjectSource[]{new ProjectSource()};

        final Gson gson = new GsonBuilder().setPrettyPrinting().create();
        try (OutputStreamWriter writer = new OutputStreamWriter(new FileOutputStream(directory.resolve(projectDef.getFileName()).toFile()))) {
            gson.toJson(projectFile, writer);
        }
    }

    private static void saveNamespace(NamespaceDef namespaceDef, Path directory) throws IOException {
        Path currentDirectory = directory;
        for (String component : namespaceDef.getNamespace().split("\\.")) {
            currentDirectory = currentDirectory.resolve(component);
            Files.createDirectories(currentDirectory);
        }

        for (FileDef fileDef : namespaceDef.getFiles().values()) {
            final CodeWriter writer = new CodeWriter(4);
            writer.writeLine(String.format("namespace %s;", namespaceDef.getNamespace()));

            if (fileDef.getImports().size() > 0) {
                writer.newLine();
                for (ImportDef importDef : fileDef.getImports()) {
                    switch (importDef.getKind()) {
                        case NAMESPACE:
                            writer.writeLine(String.format("import %s.*;", ((ImportNamespaceDef) importDef).getTarget().getNamespace()));
                            break;

                        case TYPE:
                            ImportTypeDef importTypeDef = (ImportTypeDef) importDef;
                            writer.writeLine(importTypeDef.getAlias() == null
                                ? String.format("import %s.%s;", importTypeDef.getTarget().getNamespace().getNamespace(), importTypeDef.getTarget().getName())
                                : String.format("import %s.%s as %s;", importTypeDef.getTarget().getNamespace().getNamespace(), importTypeDef.getTarget().getName(), importTypeDef.getAlias()));
                            break;

                        default:
                            throw new IllegalArgumentException("Type of import directive is unknown: " + importDef.getKind());
                    }
                }
            }

            if (fileDef.getOptions().size() > 0) {
                writer.newLine();
                for (NameValuePair option : fileDef.getOptions())
                    writer.writeLine(String.format("option %s = %s;", option.getName(), option.getValue().toString()));
            }

            saveTypeScope(fileDef, writer);

            try (OutputStreamWriter stream = new OutputStreamWriter(new FileOutputStream(currentDirectory.resolve(fileDef.getFileName()).toFile()))) {
                writer.saveTo(stream);
            }
        }
    }

    private static void saveTypeScope(TypeScope scope, CodeWriter writer) {
        for (DecoratorDef decoratorDef : scope.getDefinedDecorators()) {
            writer.newLine();
            saveDecorator(decoratorDef, writer);
        }
        for (EnumerationDef enumerationDef : scope.getDefinedEnumerations()) {
            writer.newLine();
            saveEnumeration(enumerationDef, writer);
        }
        for (InterfaceDef interfaceDef : scope.getDefinedInterfaces()) {
            writer.newLine();
            saveInterface(interfaceDef, writer);
        }
        for (ClassDef classDef : scope.getDefinedClasses()) {
            writer.newLine();
            saveClass(classDef, writer);
        }
    }

    private static void saveDecorator(DecoratorDef decoratorDef, CodeWriter writer) {
        saveComments(decoratorDef, writer);
        saveDecorators(decoratorDef, writer);

        writer.write(String.format("decorator %s {", decoratorDef.getName()));
        writer.indent();

        if (decoratorDef.getDefinedConstants().size() > 0) {
            writer.newLine();
            saveConstants(decoratorDef, writer);
        }

        saveTypeScope(decoratorDef, writer);

        if (decoratorDef.getProperties().size() > 0) {
            writer.newLine();
            for (int i = 0; i < decoratorDef.getProperties().size(); i += 1) {
                final DecoratorPropertyDef propertyDef = decoratorDef.getProperties().get(i);
                saveComments(propertyDef, writer);
                saveDecorators(propertyDef, writer);

                if (propertyDef.getDefault() == null)
                    writer.writeLine(String.format("%s %s;", propertyDef.getType().toString(), propertyDef.getName()));
                else
                    writer.writeLine(String.format("%s %s = %s;", propertyDef.getType().toString(), propertyDef.getName(), propertyDef.getDefault().toString()));
                if (i + 1 < decoratorDef.getProperties().size())
                    writer.newLine();
            }
        }

        writer.dedent();
        writer.writeLine("}");
    }

    private static String luminaryTypeName(IntegralType type) {
        switch (type) {
            case INT8:
                return "Int8";

            case UINT8:
                return "Int8";

            case INT16:
                return "Int8";

            case UINT16:
                return "Int8";

            case INT32:
                return "Int8";

            case UINT32:
                return "Int8";

            case INT64:
                return "Int8";

            case UINT64:
                return "Int8";

            default:
                throw new IllegalArgumentException("Integral type '" + type + "' is unknown.");
        }
    }

    private static void saveEnumeration(EnumerationDef enumerationDef, CodeWriter writer) {
        saveComments(enumerationDef, writer);
        saveDecorators(enumerationDef, writer);

        writer.write(String.format("enum %s : %s {", enumerationDef.getName(), luminaryTypeName(enumerationDef.getUnderlyingType())));
        writer.indent();

        if (enumerationDef.getMembers().size() > 0) {
            writer.newLine();
            for (int i = 0; i < enumerationDef.getMembers().size(); i += 1) {
                final EnumerationMemberDef memberDef = enumerationDef.getMembers().get(i);
                saveComments(memberDef, writer);
                saveDecorators(memberDef, writer);

                writer.writeLine(String.format("%s = %s;", memberDef.getName(), memberDef.getValue().toString()));
                if (i + 1 < enumerationDef.getMembers().size())
                    writer.newLine();
            }
        }

        writer.dedent();
        writer.writeLine("}");
    }

    private static void saveInterface(InterfaceDef interfaceDef, CodeWriter writer) {
        saveComments(interfaceDef, writer);
        saveDecorators(interfaceDef, writer);

        writer.write(String.format("interface %s ", interfaceDef.getName()));
        if (interfaceDef.getSuperInterfaces().size() > 0) {
            writer.write(": " + interfaceDef.getSuperInterfaces().get(0).getName());
            for (int i = 1; i < interfaceDef.getSuperInterfaces().size(); i += 1)
                writer.write(", " + interfaceDef.getSuperInterfaces().get(i).getName());
        }
        writer.write(" {");
        writer.indent();

        saveTypeScope(interfaceDef, writer);

        if (interfaceDef.getProperties().size() > 0) {
            writer.newLine();
            for (int i = 0; i < interfaceDef.getProperties().size(); i += 1) {
                final InterfacePropertyDef propertyDef = interfaceDef.getProperties().get(i);
                saveComments(propertyDef, writer);
                saveDecorators(propertyDef, writer);

                writer.write(String.format("%s %s", propertyDef.getType().toString(), propertyDef.getName()));
                if (propertyDef.isReadable())
                    writer.write(" get");
                if (propertyDef.isWritable())
                    writer.write(" set");
                writer.writeLine(";");
                if (i + 1 < interfaceDef.getProperties().size())
                    writer.newLine();
            }
        }

        writer.dedent();
        writer.writeLine("}");
    }

    private static void saveClass(ClassDef classDef, CodeWriter writer) {
        saveComments(classDef, writer);
        saveDecorators(classDef, writer);

        writer.write(String.format(classDef.isFinal() ? "final class" : "class %s", classDef.getName()));
        if (classDef.getSuperClass() != null || classDef.getSuperInterfaces().size() > 0) {
            boolean isFirst = true;
            if (classDef.getSuperClass() != null) {
                isFirst = false;
                writer.write(String.format(": %s", classDef.getSuperClass().getName()));
            }
            for (InterfaceDef superInterfaceDef : classDef.getSuperInterfaces()) {
                if (isFirst) {
                    isFirst = false;
                    writer.write(String.format(": %s", superInterfaceDef.getName()));
                } else
                    writer.write(String.format(", %s", superInterfaceDef.getName()));
            }
        }
        writer.write(" {");
        writer.indent();

        if (classDef.getDefinedConstants().size() > 0) {
            writer.newLine();
            saveConstants(classDef, writer);
        }

        saveTypeScope(classDef, writer);

        if (classDef.getProperties().size() > 0) {
            writer.newLine();
            for (int i = 0; i < classDef.getProperties().size(); i += 1) {
                final ClassPropertyDef propertyDef = classDef.getProperties().get(i);
                saveComments(propertyDef, writer);
                saveDecorators(propertyDef, writer);

                final String prefix = propertyDef.isFinal()
                    ? (propertyDef.isOverride() ? "final override " : "final ")
                    : (propertyDef.isOverride() ? "override " : "");
                if (propertyDef.getDefault() == null)
                    writer.writeLine(String.format("%s%s %s;", prefix, propertyDef.getType().toString(), propertyDef.getName()));
                else
                    writer.writeLine(String.format("%s%s %s = %s;", prefix, propertyDef.getType().toString(), propertyDef.getName(), propertyDef.getDefault().toString()));
                if (i + 1 < classDef.getProperties().size())
                    writer.newLine();
            }
        }

        writer.dedent();
        writer.writeLine("}");
    }

    private static void saveComments(ItemDef commentable, CodeWriter writer) {
        for (String comment : commentable.getComments())
            writer.writeLine("/// " + comment);
    }

    private static void saveDecorators(ItemDef decoratable, CodeWriter writer) {
        for (Decorator decorator : decoratable.getDecorators()) {
            final DecoratorDef decoratorDef = decorator.getDefinition();
            writer.write(String.format("[%s", decoratorDef.getName()));

            if (decorator.getArguments().size() > 0 && decoratorDef.getProperties().size() > 0) {
                writer.write("(");

                boolean isFirst = true;
                for (int i = 0; i < decorator.getArguments().size(); i += 1) {
                    final DecoratorPropertyDef propertyDef = decorator.getArguments().get(i).getDefinition();
                    if (!decorator.getArguments().get(i).getValue().equals(propertyDef.getDefault())) {
                        if (!isFirst)
                            writer.write(", ");
                        else
                            isFirst = false;

                        writer.write(String.format("%s = %s", propertyDef.getName(), decorator.getArguments().get(i).getValue()));
                    }
                }

                writer.write(")");
            }

            writer.writeLine("]");
        }
    }

    private static void saveConstants(ConstantScope constantScope, CodeWriter writer) {
        for (int i = 0; i < constantScope.getDefinedConstants().size(); i += 1) {
            final ConstantDef constantDef = constantScope.getDefinedConstants().get(i);
            saveComments(constantDef, writer);
            saveDecorators(constantDef, writer);
            writer.writeLine(String.format("const %s %s = %s;", constantDef.getType().toString(), constantDef.getName(), constantDef.getValue().toString()));
            if (i + 1 < constantScope.getDefinedConstants().size())
                writer.newLine();
        }
    }

    public static void main(String[] args) throws IOException {
        final ProjectLoader loader = new ProjectLoader();
        System.setProperty("luminary.search.path", "C:\\Projects\\MaxiMarkets\\ProtocolGenerators\\luminary");
        final ProjectDef projectDef = loader.load("C:\\Projects\\MaxiMarkets\\TradingServer\\luminary\\External\\External.json");
        ProjectSaver.save(projectDef, "D:\\Temporary\\Luminary");
    }
}

File: csharp/main/ProjectSaver.cs
using System;
using System.IO;
using System.Linq;
using Newtonsoft.Json;

namespace Deltix.Luminary
{
	public static class ProjectSaver
	{
		/// <summary>
		/// Saves the Luminary project (its entire content including the *.lux files) into given directory.
		/// </summary>
		/// <param name="projectDef"></param>
		/// <param name="directory"></param>
		public static void Save(ProjectDef projectDef, String directory)
		{
			if (projectDef.FileName == null)
				throw new InvalidOperationException("Built-in projects cannot be exported.");
			if (!Directory.Exists(directory))
				throw new InvalidOperationException("Directory does not exist.");

			SaveProjectFile(projectDef, directory);
			foreach (NamespaceDef namespaceDef in projectDef.Namespaces.Values)
				SaveNamespace(namespaceDef, directory);
		}

		private static void SaveProjectFile(ProjectDef projectDef, String directory)
		{
			ProjectFile projectFile = new ProjectFile
			{
				References = projectDef.References.Select(x => x.Key).ToArray(),
				Sources = new[] { new ProjectSource() }
			};

			using (StreamWriter stream = new StreamWriter(new FileStream(Path.Combine(directory, projectDef.FileName), FileMode.Create, FileAccess.ReadWrite)))
				stream.WriteLine(JsonConvert.SerializeObject(projectFile, Formatting.Indented));
		}

		private static void SaveNamespace(NamespaceDef namespaceDef, String directory)
		{
			String currentDirectory = directory;
			foreach (String component in namespaceDef.Namespace.Split('.'))
				currentDirectory = Directory.CreateDirectory(Path.Combine(currentDirectory, component)).FullName;

			foreach (FileDef fileDef in namespaceDef.Files.Values)
			{
				CodeWriter writer = new CodeWriter(4);
				writer.WriteLine($"namespace {namespaceDef.Namespace};");

				if (fileDef.Imports.Count > 0)
				{
					writer.NewLine();
					foreach (ImportDef importDef in fileDef.Imports)
					{
						switch (importDef.Kind)
						{
							case ImportKind.Namespace:
								writer.WriteLine($"import {((ImportNamespaceDef)importDef).Target}.*;");
								break;

							case ImportKind.Type:
								ImportTypeDef importTypeDef = (ImportTypeDef)importDef;
								writer.WriteLine(importTypeDef.Alias == null
									? $"import {importTypeDef.Target.Namespace.Namespace}.{importTypeDef.Target.Name};"
									: $"import {importTypeDef.Target.Namespace.Namespace}.{importTypeDef.Target.Name} as {importTypeDef.Alias};");
								break;

							default:
								throw new ArgumentOutOfRangeException();
						}
					}
				}

				if (fileDef.Options.Count > 0)
				{
					writer.NewLine();
					foreach (NameValuePair option in fileDef.Options)
						writer.WriteLine($"option {option.Name} = {option.Value};");
				}

				SaveTypeScope(fileDef, writer);

				using (StreamWriter stream = new StreamWriter(new FileStream(Path.Combine(currentDirectory, fileDef.FileName), FileMode.Create, FileAccess.ReadWrite)))
					writer.SaveTo(stream);
			}
		}

		private static void SaveTypeScope(ITypeScope scope, CodeWriter writer)
		{
			foreach (DecoratorDef decoratorDef in scope.DefinedDecorators)
			{
				writer.NewLine();
				SaveDecorator(decoratorDef, writer);
			}
			foreach (EnumerationDef enumerationDef in scope.DefinedEnumerations)
			{
				writer.NewLine();
				SaveEnumeration(enumerationDef, writer);
			}
			foreach (InterfaceDef interfaceDef in scope.DefinedInterfaces)
			{
				writer.NewLine();
				SaveInterface(interfaceDef, writer);
			}
			foreach (ClassDef decortorDef in scope.DefinedClasses)
			{
				writer.NewLine();
				SaveClass(decortorDef, writer);
			}
		}

		private static void SaveDecorator(DecoratorDef decoratorDef, CodeWriter writer)
		{
			SaveComments(decoratorDef, writer);
			SaveDecorators(decoratorDef, writer);

			writer.Write("decorator {0} {{", decoratorDef.Name);
			writer.Indent();

			if (decoratorDef.DefinedConstants.Count > 0)
			{
				writer.NewLine();
				SaveConstants(decoratorDef, writer);
			}

			SaveTypeScope(decoratorDef, writer);

			if (decoratorDef.Properties.Count > 0)
			{
				writer.NewLine();
				for (Int32 index = 0; index < decoratorDef.Properties.Count; index += 1)
				{
					DecoratorPropertyDef propertyDef = decoratorDef.Properties[index];
					SaveComments(propertyDef, writer);
					SaveDecorators(propertyDef, writer);

					if (propertyDef.Default == null)
						writer.WriteLine("{0} {1};", propertyDef.Type, propertyDef.Name);
					else
						writer.WriteLine("{0} {1} = {2};", propertyDef.Type, propertyDef.Name, propertyDef.Default);
					if (index + 1 < decoratorDef.Properties.Count)
						writer.NewLine();
				}
			}

			writer.Dedent();
			writer.WriteLine("}");
		}

		private static void SaveEnumeration(EnumerationDef enumerationDef, CodeWriter writer)
		{
			SaveComments(enumerationDef, writer);
			SaveDecorators(enumerationDef, writer);

			writer.Write("enum {0} : {1} {{", enumerationDef.Name, enumerationDef.UnderlyingType);
			writer.Indent();

			if (enumerationDef.Members.Count > 0)
			{
				writer.NewLine();
				for (Int32 index = 0; index < enumerationDef.Members.Count; index += 1)
				{
					EnumerationMemberDef memberDef = enumerationDef.Members[index];
					SaveComments(memberDef, writer);
					SaveDecorators(memberDef, writer);

					writer.WriteLine("{0} = {1};", memberDef.Name, memberDef.Value);
					if (index + 1 < enumerationDef.Members.Count)
						writer.NewLine();
				}
			}

			writer.Dedent();
			writer.WriteLine("}");
		}

		private static void SaveInterface(InterfaceDef interfaceDef, CodeWriter writer)
		{
			SaveComments(interfaceDef, writer);
			SaveDecorators(interfaceDef, writer);

			writer.Write("interface {0} ", interfaceDef.Name);
			if (interfaceDef.SuperInterfaces.Count > 0)
			{
				writer.Write(": {0}", interfaceDef.SuperInterfaces[0].Name);
				foreach (InterfaceDef superInterfaceDef in interfaceDef.SuperInterfaces.Skip(1))
					writer.Write(", {0}", superInterfaceDef.Name);
			}
			writer.Write(" {");
			writer.Indent();

			SaveTypeScope(interfaceDef, writer);

			if (interfaceDef.Properties.Count > 0)
			{
				writer.NewLine();
				for (Int32 index = 0; index < interfaceDef.Properties.Count; index += 1)
				{
					InterfacePropertyDef propertyDef = interfaceDef.Properties[index];
					SaveComments(propertyDef, writer);
					SaveDecorators(propertyDef, writer);

					if (propertyDef.IsOverride)
						writer.Write("override ");
					writer.Write("{0} {1}", propertyDef.Type, propertyDef.Name);
					if (propertyDef.IsReadable)
						writer.Write(" get");
					if (propertyDef.IsWritable)
						writer.Write(" set");
					writer.WriteLine(";");
					if (index + 1 < interfaceDef.Properties.Count)
						writer.NewLine();
				}
			}

			writer.Dedent();
			writer.WriteLine("}");
		}

		private static void SaveClass(ClassDef classDef, CodeWriter writer)
		{
			SaveComments(classDef, writer);
			SaveDecorators(classDef, writer);

			writer.Write(classDef.IsFinal ? "final class {0}" : "class {0}", classDef.Name);
			if (classDef.SuperClass != null || classDef.SuperInterfaces.Count > 0)
			{
				Boolean isFirst = true;
				if (classDef.SuperClass != null)
				{
					isFirst = false;
					writer.Write(": {0}", classDef.SuperClass.Name);
				}
				foreach (InterfaceDef superInterfaceDef in classDef.SuperInterfaces)
				{
					if (isFirst)
					{
						isFirst = false;
						writer.Write(": {0}", superInterfaceDef.Name);
					}
					else
						writer.Write(", {0}", superInterfaceDef.Name);
				}
			}
			writer.Write(" {");
			writer.Indent();

			if (classDef.DefinedConstants.Count > 0)
			{
				writer.NewLine();
				SaveConstants(classDef, writer);
			}

			SaveTypeScope(classDef, writer);

			if (classDef.Properties.Count > 0)
			{
				writer.NewLine();
				for (Int32 index = 0; index < classDef.Properties.Count; index += 1)
				{
					ClassPropertyDef propertyDef = classDef.Properties[index];
					SaveComments(propertyDef, writer);
					SaveDecorators(propertyDef, writer);

					String prefix = propertyDef.IsFinal
						? (propertyDef.IsOverride ? "final override " : "final ")
						: (propertyDef.IsOverride ? "override ": "");
					if (propertyDef.Default == null)
						writer.WriteLine("{0}{1} {2};", prefix, propertyDef.Type, propertyDef.Name);
					else
						writer.WriteLine("{0}{1} {2} = {3};", prefix, propertyDef.Type, propertyDef.Name, propertyDef.Default);
					if (index + 1 < classDef.Properties.Count)
						writer.NewLine();
				}
			}

			writer.Dedent();
			writer.WriteLine("}");
		}

		private static void SaveComments(ICommentable commentable, CodeWriter writer)
		{
			foreach (String comment in commentable.Comments)
				writer.WriteLine("/// {0}", comment);
		}

		private static void SaveDecorators(IDecoratable decoratable, CodeWriter writer)
		{
			foreach (Decorator decorator in decoratable.Decorators)
			{
				DecoratorDef decoratorDef = decorator.Definition;
				writer.Write("[{0}", decoratorDef.Name);

				if (decorator.Arguments.Count > 0 && decoratorDef.Properties.Count > 0)
				{
					writer.Write("(");

					Boolean isFirst = true;
					for (Int32 i = 0; i < decorator.Arguments.Count; i += 1)
					{
						DecoratorPropertyDef propertyDef = decorator.Arguments[i].Definition;
						if (!decorator.Arguments[i].Value.Equals(propertyDef.Default))
						{
							if (!isFirst)
								writer.Write(", ");
							else
								isFirst = false;

							writer.Write("{0} = {1}", propertyDef.Name, decorator.Arguments[i].Value);
						}
					}

					writer.Write(")");
				}
				writer.WriteLine("]");
			}
		}

		private static void SaveConstants(IConstantScope constantScope, CodeWriter writer)
		{
			for (Int32 index = 0; index < constantScope.DefinedConstants.Count; index += 1)
			{
				ConstantDef constantDef = constantScope.DefinedConstants[index];
				SaveComments(constantDef, writer);
				SaveDecorators(constantDef, writer);
				writer.WriteLine("const {0} {1} = {2};", constantDef.Type, constantDef.Name, constantDef.Value);
				if (index + 1 < constantScope.DefinedConstants.Count)
					writer.NewLine();
			}
		}
	}
}
